grakn-graql/src/main/java/ai/grakn/graql/internal/analytics/ShortestPathVertexProgram.java

```diff
@@ -177,59 +176,138 @@ public void safeExecute(final Vertex vertex, Messenger<Tuple> messenger, final M
+                        hasMessageDestination = true;
+                        predecessorFromDestination = (String) message.getValue(0);
+                        messenger.sendMessage(messageScopeIn, Pair.with(id, -1));
+                        messenger.sendMessage(messageScopeOut, Pair.with(id, -2));
+                        vertex.property(PREDECESSOR, predecessorFromDestination);
+                        memory.and(VOTE_TO_HALT_DESTINATION, false);
+                    }
+                }
+                if (hasMessageSource && hasMessageDestination) {
+                    LOGGER.debug("Found path");
+                    memory.or(FOUND_PATH, true);
+                    memory.set(PREDECESSOR_FROM_SOURCE, predecessorFromSource);
+                    memory.set(PREDECESSOR_FROM_DESTINATION, predecessorFromDestination);
 
-            if (persistentProperties.get(END_ID).equals(id)) {
-                memory.or(FOUND_DESTINATION, true);
-                memory.set(NEXT_DESTINATION, fromVertex);
-                LOGGER.debug("Found the destination vertex");
-                return;
+                    vertex.property(FOUND_IN_ITERATION, memory.getIteration());
+                    return;
+                }
             }
-
-            vertex.property(FROM_VERTEX, fromVertex);
-            messenger.sendMessage(messageScopeIn, Pair.with(id, 1));
-            messenger.sendMessage(messageScopeOut, Pair.with(id, 2));
-            LOGGER.debug("FromVertex added, messages sent");
-
-            memory.and(VOTE_TO_HALT, false);
         }
     }
 
     private void updateCasting(Vertex vertex, Messenger<Tuple> messenger, Memory memory) {
-        LOGGER.debug("Considering vertex " + vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name()));
-        boolean messageSentIn = false;
-        boolean messageSentOut = false;
-        Iterator<Tuple> receiveMessages = messenger.receiveMessages();
-        while (receiveMessages.hasNext()) {
-            Tuple message = receiveMessages.next();
-            if ((int) message.getValue(1) == 1) {
-                if (!messageSentIn) {
-                    messenger.sendMessage(messageScopeIn, message);
-                    messageSentIn = true;
-                }
+        LOGGER.debug("Considering casting " + vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name()));
+        Map<Integer, Tuple> messageMap = new HashMap<>();
+        Iterator<Tuple> iterator = messenger.receiveMessages();
+        int i = 0;
+        while (iterator.hasNext()) {
+            Tuple message = iterator.next();
+            LOGGER.debug("Message " + i++ + ": " + message.getValue(0));
+            messageMap.put((int) message.getValue(1), message);
+        }
+        sendMessagesFromCasting(messenger, memory, messageMap);
+    }
+
+    private void sendMessagesFromCasting(Messenger<Tuple> messenger, Memory memory, Map<Integer, Tuple> messageMap) {
```

Comment:
I have no idea what this method is doing. Some java doccing for huge complex methods like this would be very helpful.